// @sap/ux-specification/dist/src/sync/v4/utils/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../../common/utils");
const StableIdHelper_1 = require("./StableIdHelper");
/**
 * Adds the line item records to the app schema (for the list report or an object page section comprising a table)
 * @param appSchema - the app specific schema that shall get enhanced
 * @param lineItemAnnotation - the UI.LineItem annotation, comprising all records
 * @param entityType - the entity type as part of the AVT ConverterOutput
 * @param lineItemId - optional: in case of OP the paramter must be passed to distinguish the OP tables;
 *                      in case of LR or ALP no ID is passed, 'LineItems' is taken then.
 */
function addLineItemsType(appSchema, lineItemAnnotation, entityType, lineItemId) {
    let schemaKey;
    //ToDo: by the next measure the custom columns are eliminated, for the time being, from the schema
    appSchema['definitions']['TableColumn'] = JSON.parse(JSON.stringify(appSchema['definitions']['GenericColumns']['additionalProperties'].anyOf[0]));
    lineItemId = lineItemId || 'LineItems';
    appSchema['definitions'][lineItemId] = {
        type: 'object',
        properties: {},
        additionalProperties: false
    };
    if (lineItemAnnotation) {
        let i = 0;
        lineItemAnnotation.forEach((lineItemRecord) => {
            schemaKey = StableIdHelper_1.getStableIdPartFromDataField(lineItemRecord);
            if (schemaKey) {
                appSchema['definitions'][lineItemId]['properties'][schemaKey] = {
                    $ref: '#/definitions/TableColumn',
                    description: utils_1.getColumnDescription(lineItemRecord, entityType),
                    // Custom property in schema - for object properties ordering purpose
                    propertyIndex: i
                };
                i++;
            }
        });
    }
    //ToDo, yet out of scope: add custom columns
    //appSchema['definitions']['custom'] = appSchema['definitions']['GenericColumns']['additionalProperties'].anyOf[1];
}
exports.addLineItemsType = addLineItemsType;
