// @sap/ux-specification/dist/src/sync/v4/export/export.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const pages_1 = require("./pages");
const controls_1 = require("./controls");
const common_1 = require("../../common");
const application_1 = require("../application");
const manifest_1 = require("./manifest");
const flexibleColumnLayout_1 = require("./flexibleColumnLayout");
const utils_1 = require("../../common/utils");
const common_2 = require("../../../specification/common");
const TableColumn_1 = require("./controls/TableColumn");
/**
 * Return the skeleton of a exportResultManifest's "sap.ui.generic.app" entry, prefilled with some values
 * @param name - name of the template, e.g. "sap.suite.ui.generic.template.ObjectPage"
 * @param entitySet - the main entity set
 * @param navigationProperty - navigation property
 */
const getV4ManifestPageTemplate = (name, entitySet) => {
    const V4Page = {
        type: 'Component',
        id: '',
        name: name,
        options: {
            settings: {
                entitySet: entitySet,
                navigation: {},
                controlConfiguration: {}
            }
        }
    };
    return V4Page;
};
const defaultExportResult = {
    flexChanges: [],
    manifest: {
        'sap.ui5': {
            flexEnabled: true,
            routing: {
                targets: {}
            }
        },
        'sap.app': {}
    },
    fragments: []
};
/**
 * Evaluates an export rule for a property or object, transfers to flex change or manifest setting
 * @param syncRule - export rule from the object classes decorator
 * @param baseId - selector id of the current page
 * @param breadcrumbs - list of breadcrumbs, i.e. ID parts for building the stable ID
 * @param localIdParts - list of parent and child ID's
 * @param key - key of the given property
 * @param configObject - current (sub)object of the configuration file
 * @param exportResults - object with all collected manifest entries and flex changes
 * @param fullManifest - old or existing manifest
 * @param ids - list of parent & child ids
 * @param title - title from the app schema (comprising for instance the facet ID)
 * @param pageName - array comprising the page name
 */
function evaluateExportRule(syncRule, baseId, breadcrumbs, key, configObject, exportResults, fullManifest, ids, title, pageName, ui5Version) {
    // handle flex change
    if (syncRule.flex) {
        const flexChange = common_1.createFlexChange(configObject, breadcrumbs, syncRule, ui5Version, baseId, ids, title, key);
        exportResults.flexChanges.push(syncRule.flex.exportFunction(flexChange, fullManifest));
    }
    else if (syncRule.manifest) {
        // handle manifest change
        //eliminate subsections' hierarchy
        let index = ids.indexOf('subsections');
        while (index > -1) {
            ids.splice(index - 1, 2);
            index = ids.indexOf('subsections');
        }
        let targetAnnotation = '', targetAnnotationEncoded = '', columnKey, custom = false;
        if (ids[0] === 'sections') {
            if (ids[1] === 'custom' && ids[2] !== ids[ids.length - 1]) {
                custom = true;
                targetAnnotationEncoded = ids[2];
                targetAnnotation = targetAnnotationEncoded.replace(/::/g, '/');
            }
            else {
                targetAnnotationEncoded = ids[1] && ids[1].replace(/\//g, '::');
                targetAnnotation = targetAnnotationEncoded.replace(/::/g, '/');
            }
        }
        if (ids[ids.length - 3] === 'columns') {
            columnKey = ids[ids.length - 2];
        }
        const path = syncRule.manifest.path(pageName, {}, targetAnnotationEncoded, columnKey);
        const manifestSection = common_1.getManifestSectionByPathV4(exportResults.manifest, path, targetAnnotation, custom ? undefined : targetAnnotationEncoded);
        const manifestKey = syncRule.manifest.key || key;
        if (configObject[key] !== undefined) {
            const exportHandler = syncRule.manifest.export;
            if (exportHandler !== false) {
                if (exportHandler && typeof exportHandler === 'function') {
                    exportHandler(manifestSection, configObject, key);
                }
                else {
                    manifestSection[manifestKey] = configObject[key];
                }
            }
        }
        else if (manifestSection[manifestKey] !== undefined) {
            delete manifestSection[manifestKey];
        }
        utils_1.deleteEmptyStructure(exportResults.manifest, path, targetAnnotation, targetAnnotationEncoded);
    }
}
/**
 * Recursive function that traverses the content of the current object. It reads decorators and fills exportResultManifest entries and flex changes
 * @param configObject - current object to traverse
 * @param parentIds - array of collected parent ids
 * @param stableIdParts - list of parts of the stable ID for flex changes
 * @param breadcrumbs - array of sequence of properties that we are currently processing. Required for instance for table column name
 * @param exportResults - result object with all collected exportResultManifest entries and flex changes
 * @param appId - id of the Fiori elements app, which is usually namespace.appid
 * @param baseId - selector id of the current page
 * @param pageName - array comprising the page name
 * @param jsonSchema - application specific schema
 * @param schemaDefinition - the current entry point or definition in the app schema
 */
function transferManifestEntriesAndFlexChange(fullManifest, configObject, parentIds, stableIdParts, breadcrumbs, exportResults, appId, baseId, pageName, jsonSchema, schemaDefinition, ui5Version, title) {
    let currentConfigObject;
    if (schemaDefinition && schemaDefinition['properties']) {
        for (const key in schemaDefinition['properties']) {
            currentConfigObject = configObject[key];
            const childId = common_1.getChildId(configObject, key);
            const ids = childId ? [...parentIds, childId] : [...parentIds, key];
            const localIdParts = childId ? [...stableIdParts, childId] : stableIdParts;
            const syncRule = common_1.getReflectMetadata(configObject, key);
            if (syncRule) {
                evaluateExportRule(syncRule, baseId, breadcrumbs, key, configObject, exportResults, fullManifest, ids, title, pageName, ui5Version);
                continue;
            }
            const propertyDefinition = schemaDefinition['properties'][key];
            if (typeof currentConfigObject === 'object' && propertyDefinition) {
                let objects = [], isArray = false, definitionArray = [];
                if (propertyDefinition.$ref) {
                    objects.push(currentConfigObject);
                    definitionArray = propertyDefinition.$ref.split('#/definitions/');
                }
                else if (propertyDefinition['anyOf']) {
                    objects.push(currentConfigObject);
                    propertyDefinition['anyOf'].forEach((element) => {
                        if (element.$ref.includes(currentConfigObject.constructor.name)) {
                            definitionArray.push(...element.$ref.split('#/definitions/'));
                        }
                    });
                    if (definitionArray.length === 0) {
                        definitionArray = propertyDefinition['anyOf'][0].$ref.split('#/definitions/');
                    }
                }
                else if (propertyDefinition.type === 'array') {
                    definitionArray = propertyDefinition.items.$ref.split('#/definitions/');
                    objects = currentConfigObject;
                    isArray = true;
                }
                else {
                    objects.push(currentConfigObject);
                }
                const nextDefinition = utils_1.getNextTargetDefinition(definitionArray, jsonSchema, title, currentConfigObject, propertyDefinition, key);
                for (const innerObject of objects) {
                    const additionalIds = [];
                    if (breadcrumbs[0] === 'sections' && key === 'custom' && isArray) {
                        // We need add custom section 'id' into breadcrumbs to retrieve correct path to manifest
                        additionalIds.push(innerObject['id']);
                    }
                    transferManifestEntriesAndFlexChange(fullManifest, innerObject, [...ids, ...additionalIds], localIdParts, [...breadcrumbs, key], exportResults, appId, baseId, pageName, jsonSchema, nextDefinition.targetDefinition, ui5Version, nextDefinition.title);
                }
            }
        }
    }
}
/**
 * Instantiates the export classes for the object page section
 * @param objectPageConfig - the given object page configuration
 * @param objectPage - the overall object page export class where the sections shall be included
 */
function getExportClassesForSections(objectPageConfig, objectPage) {
    /**
     * Instantiates the export classes for a single object page section
     * @param section - Export class as an object, to be updated
     * @param sectionId - Current section ID
     */
    function getClassesForSingleSection(section, sectionId) {
        if (section && section['table']) {
            section['table'] = Object.assign(new controls_1.ObjectPageTable(), section['table']);
            for (const columnKey in section['table'].columns) {
                section['table'].columns[columnKey] = Object.assign(new controls_1.ObjectPageTableColumn(), section['table'].columns[columnKey]);
            }
        }
        else if (sectionId === 'custom' && Array.isArray(objectPageConfig.sections[sectionId])) {
            const customSections = objectPageConfig.sections.custom;
            for (const index in customSections) {
                customSections[index] = Object.assign(new controls_1.ObjectPageCustomSectionFragment(), customSections[index]);
            }
        }
    }
    /**
     * Recursive sub-function for handling subsections
     * @param section - Object page section in config
     */
    function getSubsectionClasses(section) {
        if (section['subsections']) {
            let subSection;
            for (const subSectionId in section['subsections']) {
                subSection = section['subsections'][subSectionId];
                getClassesForSingleSection(subSection, subSectionId);
                getSubsectionClasses(subSection);
            }
        }
    }
    // Main
    let section;
    for (const sectionId in objectPageConfig.sections) {
        section = objectPage.sections[sectionId];
        getClassesForSingleSection(section, sectionId);
        getSubsectionClasses(section);
    }
}
/**
 * Run through the given ObjectPage config and return respective exportResultManifest entry and flex changes
 * @param appId - id of the Fiori elements app, which is usually namespace.appid
 * @param objectPageConfig - content of the src/ObjectPage_<entity_set>.json file
 * @param manifest - manifest of the application
 * @param pageName - routing target defined in manfiest
 */
const exportObjectPage = (appId, objectPageConfig, manifest, pageName, jsonSchema, ui5Version) => {
    const exportResults = defaultExportResult;
    exportResults.manifest = JSON.parse(JSON.stringify(manifest));
    const objectPage = Object.assign(new pages_1.ObjectPage(), objectPageConfig);
    objectPage.header = Object.assign(new controls_1.ObjectPageHeader(), objectPage.header);
    objectPage.layout = Object.assign(new controls_1.ObjectPageLayout(), objectPage.layout);
    const pageLayoutInformation = common_1.getPageLayoutInformation(objectPage);
    const baseId = `${appId}::${pageLayoutInformation.id}::${pageName[0]}--`;
    if (objectPageConfig && objectPageConfig.sections) {
        getExportClassesForSections(objectPageConfig, objectPage);
    }
    transferManifestEntriesAndFlexChange(exportResults.manifest, objectPage, [], //no parentIds
    [], // no stableIds
    [], // no breadcrumbs
    exportResults, appId, baseId, pageName, jsonSchema, jsonSchema, //starting point for target definition
    ui5Version);
    return exportResults;
};
/**
 * Instantiates the relevant object classes for the export of table columns.
 * Common function for LR + ALP
 * @param listReport - list report or ALP config (instantiation of table columns is the same for both)
 * @param complexIdMatch - regex for machting column IDs, diferent for LR + ALP
 */
function instantiateTableColumns(listReport) {
    if (listReport.table.columns) {
        for (const columnId in listReport.table.columns) {
            if (columnId === 'custom') {
                //ToDo
            }
            else {
                listReport.table.columns[columnId] = Object.assign(new TableColumn_1.TableColumn(), listReport.table.columns[columnId]);
            }
        }
    }
}
/**
 * Run through the given List Page (LR or ALP) config and return respective exportResultManifest entry and flex changes
 * @param appId - id of the Fiori elements app, which is usually namespace.appid
 * @param analyticalListPageConfig - content of the Analytical List page config file
 * @param pageName - routing target defined in manifest
 */
const exportListPage = (appId, listPage, manifest, pageName, jsonSchema, ui5Version) => {
    const exportResults = defaultExportResult;
    exportResults.manifest = JSON.parse(JSON.stringify(manifest));
    // when assigning a JSON structure to an object structure we need to assign all children, otherwise decorators will not exist
    // find a better way to do this generically, perhaps using the schema?
    listPage.table = Object.assign(new controls_1.Table(), listPage.table);
    instantiateTableColumns(listPage);
    const pageLayoutInformation = common_1.getPageLayoutInformation(listPage);
    const baseId = `${appId}::${pageLayoutInformation.id}::${pageName[0]}--`;
    if (!exportResults.manifest['sap.ui5']['routing']['targets'][pageName[0]]) {
        const newV4Page = getV4ManifestPageTemplate(pageLayoutInformation.pageId, pageName[0]);
        exportResults.manifest['sap.ui5']['routing']['targets'][pageName[0]] = newV4Page;
    }
    transferManifestEntriesAndFlexChange(exportResults.manifest, listPage, [], //no parentIds
    [], // no stableIds
    [], // no breadcrumbs
    exportResults, appId, baseId, pageName, jsonSchema, jsonSchema, //starting point for target definition
    ui5Version);
    return exportResults;
};
/**
 * Run through the given ListReport config and return respective exportResultManifest entry and flex changes
 * @param appId - id of the Fiori elements app, which is usually namespace.appid
 * @param listReportConfig - content of the src/ListReport_<entity_set>.json file
 * @param pageName - routing target defined in manifest
 */
const exportListReportPage = (appId, listReportConfig, manifest, pageName, jsonSchema, ui5Version) => {
    const listReport = Object.assign(new pages_1.ListReport(), listReportConfig);
    return exportListPage(appId, listReport, manifest, pageName, jsonSchema, ui5Version);
};
/**
 * Run through the given Analytical List Page config and return respective exportResultManifest entry and flex changes
 * @param appId - id of the Fiori elements app, which is usually namespace.appid
 * @param analyticalListPageConfig - content of the Analytical List page config file
 * @param pageName - routing target defined in manifest
 */
const exportAnalyticalListPage = (appId, analyticalListPageConfig, manifest, pageName, jsonSchema, ui5Version) => {
    const analyticalListPage = Object.assign(new pages_1.AnalyticalListPage(), analyticalListPageConfig);
    return exportListPage(appId, analyticalListPage, manifest, pageName, jsonSchema, ui5Version);
};
/**
 * API for the export of a V4 app configuration
 * Export means the transfer of the properties and values of the given config to manifest entries or flex changes
 * @param application - V4 application (app.json config file)
 * @param manifest - Manifest of the given app
 * @param jsonSchema - App schema
 *
 * @returns ExportResults = object comprising the updated manifest and a list of flex changes
 */
const exportApplicationV4 = (application, manifest, jsonSchema) => {
    const exportResults = defaultExportResult;
    exportResults.manifest = JSON.parse(JSON.stringify(manifest));
    const applicationV4 = Object.assign(new application_1.ApplicationV4(), application);
    //transform application settings
    applicationV4.settings = Object.assign(new application_1.AppSettings(), applicationV4.settings);
    for (const key in jsonSchema['definitions']['AppSettings']['properties']) {
        const syncRule = common_1.getReflectMetadata(applicationV4.settings, key);
        if (syncRule && syncRule.manifest) {
            const path = syncRule.manifest.path();
            const manifestSection = common_1.getManifestSectionByPathV4(exportResults.manifest, path);
            if (application['settings'] && application['settings'][key]) {
                manifestSection[key] = application['settings'][key];
            }
            else {
                delete manifestSection[key];
            }
            utils_1.deleteEmptyStructure(exportResults.manifest, path);
        }
    }
    //transform routing
    exportResults.manifest['sap.ui5'].routing = manifest_1.transformRoutingV4(application['home'], application['pages'], exportResults.manifest);
    // Update FCL data
    flexibleColumnLayout_1.updateFcl(exportResults.manifest['sap.ui5'], application);
    return exportResults;
};
/**
 * General API for the export of a V4 config page
 * Export means the transfer of the properties and values of the given config to manifest entries or flex changes
 * @param appId - Application ID
 * @param page - Current page (config)
 * @param manifest - Manifest of the given app
 * @param jsonSchema - App schema
 *
 * @returns ExportResults = object comprising the updated manifest and a list of flex changes
 */
exports.exportPageV4 = (exportParametersV4, ui5Version) => {
    if (exportParametersV4[common_2.SchemaType.ListReport]) {
        const { appId, page, manifest, jsonSchema } = exportParametersV4[common_2.SchemaType.ListReport];
        return exportListReportPage(appId, page.config, manifest, [page.name], jsonSchema, ui5Version);
    }
    else if (exportParametersV4[common_2.SchemaType.ObjectPage]) {
        const { appId, page, manifest, jsonSchema } = exportParametersV4[common_2.SchemaType.ObjectPage];
        return exportObjectPage(appId, page.config, manifest, [page.name], jsonSchema, ui5Version);
    }
    else if (exportParametersV4[common_2.SchemaType.Application]) {
        const { application, manifest, jsonSchema } = exportParametersV4[common_2.SchemaType.Application];
        return exportApplicationV4(application, manifest, jsonSchema);
    }
    else if (exportParametersV4[common_2.SchemaType.AnalyticalListPage]) {
        const { appId, page, manifest, jsonSchema } = exportParametersV4[common_2.SchemaType.AnalyticalListPage];
        return exportAnalyticalListPage(appId, page.config, manifest, [page.name], jsonSchema, ui5Version);
    }
};
