// @sap/ux-specification/dist/src/sync/v2/import/pages/analyticalListPage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("../common/index");
const page_1 = require("../../../../specification/common/page");
const controls_1 = require("../../export/controls");
const AnalyticalListPageChart_1 = require("../../../v2/export/controls/AnalyticalListPageChart");
const AnalyticalListPageKPI_1 = require("../../../v2/export/controls/AnalyticalListPageKPI");
const AnalyticalListPage_1 = require("../../export/pages/AnalyticalListPage");
const extensionLogger_1 = require("../../../../extensionLogger");
const i18n_1 = require("../../../../i18n/i18n");
const i18next_1 = __importDefault(require("i18next"));
const utils_1 = require("../../import/utils");
const listReport_1 = require("../pages/listReport");
function addSettings(analyticalListPageConfig, manifest, jsonSchema, pageKey, logger) {
    // Check table settings from target structure
    analyticalListPageConfig.table = Object.assign(new controls_1.AnalyticalTable(), analyticalListPageConfig.table);
    index_1.transferSettingsOfObject(analyticalListPageConfig.table, manifest, jsonSchema['definitions']['AnalyticalTableALP<LineItems>'], pageKey, logger);
    analyticalListPageConfig.filterBar = Object.assign(new controls_1.AnalyticalListPageFilterBar(), analyticalListPageConfig.filterBar);
    index_1.transferSettingsOfObject(analyticalListPageConfig.filterBar, manifest, jsonSchema['definitions']['AnalyticalListPageFilterBar'], pageKey, logger);
    analyticalListPageConfig.chart = Object.assign(new AnalyticalListPageChart_1.ChartSettings(), analyticalListPageConfig.chart);
    index_1.transferSettingsOfObject(analyticalListPageConfig.chart, manifest, jsonSchema['definitions']['ChartSettings'], pageKey, logger);
    analyticalListPageConfig.keyPerformanceIndicators = {};
    const kpiCards = manifest['sap.ui.generic.app']['pages'][pageKey].component.settings.keyPerformanceIndicators;
    if (kpiCards) {
        Object.keys(kpiCards).forEach((kpiKey) => {
            analyticalListPageConfig.keyPerformanceIndicators[kpiKey] = new AnalyticalListPageKPI_1.KPISettings();
            analyticalListPageConfig.keyPerformanceIndicators = Object.assign(new AnalyticalListPageKPI_1.KPISettings(), analyticalListPageConfig.keyPerformanceIndicators);
            index_1.transferSettingsOfObject(analyticalListPageConfig.keyPerformanceIndicators[kpiKey], manifest, jsonSchema['definitions']['KPISettings'], pageKey, logger, kpiKey);
        });
    }
}
function createAnalyticalListPageConfig(manifest, flex, appSchema, fragments, logger) {
    // Initialize i18next
    i18n_1.initI18n();
    if (!manifest['sap.ui.generic.app']) {
        extensionLogger_1.log(logger, {
            severity: "error" /* Error */,
            message: i18next_1.default.t('NOFE'),
            location: {
                path: 'webapp/manifest.json',
                range: ['sap.ui.generic.app']
            }
        });
        return;
    }
    const v2Pages = manifest['sap.ui.generic.app'].pages;
    if (!v2Pages) {
        extensionLogger_1.log(logger, {
            severity: "error" /* Error */,
            message: i18next_1.default.t('NOPAGES', { appId: manifest['sap.app']['id'] }),
            location: {
                path: 'webapp/manifest.json',
                range: ['sap.ui.generic.app']
            }
        });
        return;
    }
    const config = new AnalyticalListPage_1.AnalyticalListPage();
    const pageKeys = [];
    const v2Page = utils_1.findAnalyticalListPage(manifest['sap.ui.generic.app'].pages, pageKeys);
    if (!v2Page) {
        extensionLogger_1.log(logger, {
            severity: "error" /* Error */,
            message: i18next_1.default.t('NOALP'),
            location: {
                path: 'webapp/manifest.json',
                range: ['sap.ui.generic.app']
            }
        });
        return;
    }
    addSettings(config, manifest, appSchema, pageKeys[0], logger);
    flex.forEach((change) => {
        index_1.addFlex(config, change, appSchema);
    });
    listReport_1.addFragments(config, manifest, v2Page, page_1.PageType.AnalyticalListPage, fragments);
    return config;
}
exports.createAnalyticalListPageConfig = createAnalyticalListPageConfig;
