// @sap/ux-specification/dist/src/sync/v2/import/pages/listReport.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("../common/index");
const common_1 = require("../../../../specification/common");
const controls_1 = require("../../export/controls");
const common_2 = require("../../../common");
const extensionLogger_1 = require("../../../../extensionLogger");
const i18next_1 = __importDefault(require("i18next"));
const i18n_1 = require("../../../../i18n/i18n");
const utils_1 = require("../../import/utils");
const xml_js_1 = require("xml-js");
const controls_2 = require("../../export/controls");
function addSettings(listReportConfig, manifest, jsonSchema, pageKey, logger) {
    // Check table settings from target structure
    const tableType = index_1.determineTableType(manifest, pageKey);
    let localJsonSchema;
    switch (tableType) {
        case 'TreeTable':
            listReportConfig.table = Object.assign(new controls_1.TreeTable(), listReportConfig.table);
            index_1.transferSettingsOfObject(listReportConfig.table, manifest, jsonSchema['definitions']['TreeTable<LineItems>'], pageKey, logger);
            break;
        case 'AnalyticalTable':
            listReportConfig.table = Object.assign(new controls_1.AnalyticalTable(), listReportConfig.table);
            index_1.transferSettingsOfObject(listReportConfig.table, manifest, jsonSchema['definitions']['AnalyticalTable<LineItems>'], pageKey, logger);
            break;
        case 'GridTable':
            listReportConfig.table = Object.assign(new controls_1.GridTable(), listReportConfig.table);
            index_1.transferSettingsOfObject(listReportConfig.table, manifest, jsonSchema['definitions']['GridTable<LineItems>'], pageKey, logger);
            break;
        case 'ResponsiveTable':
        default:
            listReportConfig.table = Object.assign(new controls_1.ResponsiveTable(), listReportConfig.table);
            if (jsonSchema['definitions']['ResponsiveTableWithMultiSelect<LineItems>'] &&
                jsonSchema['definitions']['ResponsiveTableWithInlineDelete<LineItems>']) {
                localJsonSchema = JSON.parse(JSON.stringify(jsonSchema['definitions']['ResponsiveTableWithMultiSelect<LineItems>']));
                localJsonSchema['properties'] = Object.assign(Object.assign({}, localJsonSchema['properties']), jsonSchema['definitions']['ResponsiveTableWithInlineDelete<LineItems>']['properties']);
                index_1.transferSettingsOfObject(listReportConfig.table, manifest, localJsonSchema, pageKey, logger);
            }
            break;
    }
    listReportConfig.filterBar = Object.assign(new controls_1.FilterBar(), listReportConfig.filterBar);
    index_1.transferSettingsOfObject(listReportConfig.filterBar, manifest, jsonSchema['definitions']['FilterBar'], pageKey, logger);
}
function evaluateFragmentColumn(column, viewExtensionsOfPage, manifestKey, keyPart3, config, extensionType, cellsFragmentName) {
    if (column['customData'] || column['table:customData']) {
        const p13nData = column['customData']
            ? column['customData']['core:CustomData']._attributes.value
            : column['table:customData']['core:CustomData']._attributes.value;
        const columnParts = JSON.parse(p13nData.substr(1, p13nData.length));
        let columnText = column['Text'] && column['Text']._attributes && column['Text']._attributes['text'];
        if (!columnText) {
            columnText = column['Label'] && column['Label']._attributes && column['Label']._attributes['text'];
        }
        const customElement = {
            id: column._attributes && column._attributes['id'],
            text: columnText,
            columnKey: columnParts['columnKey'],
            columnIndex: columnParts['columnIndex'] !== undefined ? Number(columnParts['columnIndex']) : undefined,
            leadingProperty: columnParts['leadingProperty'],
            className: viewExtensionsOfPage[manifestKey].className,
            fragmentName: viewExtensionsOfPage[manifestKey].fragmentName,
            tabKey: keyPart3,
            extensionType
        };
        if (cellsFragmentName) {
            customElement.cellsFragmentName = cellsFragmentName;
        }
        if (!config.table['columns']['custom']) {
            config.table['columns']['custom'] = [];
        }
        const customColumns = config.table['columns']['custom'];
        customColumns.push(customElement);
    }
}
function addFragments(config, manifest, v2Page, pageType, fragments, objectPageSectionData) {
    if (!fragments) {
        return;
    }
    const targetKeyEntitySet = '|' + (objectPageSectionData && objectPageSectionData.target ? objectPageSectionData.target : v2Page.entitySet);
    if (manifest['sap.ui5'] &&
        manifest['sap.ui5']['extends'] &&
        manifest['sap.ui5']['extends']['extensions'] &&
        manifest['sap.ui5']['extends']['extensions']['sap.ui.viewExtensions']) {
        const viewExtension = common_1.PAGETYPE_VIEW_EXTENSION_TEMPLATE_MAP.get(pageType);
        const viewExtensionsOfPage = manifest['sap.ui5']['extends']['extensions']['sap.ui.viewExtensions'][viewExtension];
        if (viewExtensionsOfPage) {
            for (const manifestKey of Object.keys(viewExtensionsOfPage)) {
                const manifestKeyParts = manifestKey.split('|');
                const [keyExtensionType, keyEntitySet] = manifestKeyParts;
                // If Object Page section have key, then Ignore it as it is not from QuickVariantSelectionX
                const referenceId = manifestKeyParts[2];
                const keyTab = objectPageSectionData && referenceId === objectPageSectionData.key ? undefined : referenceId;
                const extensionType = common_2.ViewTemplateType[keyExtensionType];
                if (extensionType && '|' + keyEntitySet === targetKeyEntitySet) {
                    const fragmentType = viewExtensionsOfPage[manifestKey].type
                        ? viewExtensionsOfPage[manifestKey].type.toLowerCase()
                        : 'xml';
                    let fragmentName = viewExtensionsOfPage[manifestKey].fragmentName;
                    if (!fragmentName) {
                        continue;
                    }
                    // Responsive Table ==> check cells' fragment name
                    let cellsFragmentName;
                    if (extensionType === common_2.ViewTemplateType.ResponsiveTableColumnsExtension) {
                        for (const cellsKey of Object.keys(viewExtensionsOfPage)) {
                            const [cellsKeyType, cellsKeyEntitySet, cellsKeyTab] = cellsKey.split('|');
                            if (cellsKeyType === common_2.ViewTemplateType.ResponsiveTableCellsExtension &&
                                cellsKeyEntitySet === keyEntitySet &&
                                cellsKeyTab === referenceId) {
                                cellsFragmentName = viewExtensionsOfPage[cellsKey].fragmentName;
                            }
                        }
                    }
                    // Retrieve information from fragment
                    const fragmentParts = fragmentName.split('.');
                    fragmentName = fragmentParts[fragmentParts.length - 1] + '.fragment.' + fragmentType;
                    const filteredFragments = fragments.filter((fragment) => fragment['dataSourceUri'].endsWith(fragmentName));
                    if (filteredFragments[0]) {
                        const fragmentJson = JSON.parse(xml_js_1.xml2json(filteredFragments[0].fileContent, { compact: true }));
                        const fragmentDefintion = fragmentJson[controls_2.FRAGMENT_DEFINITION];
                        const columnDefinition = controls_2.getRootCustomColumnFromFragmentJSON(fragmentDefintion);
                        if (columnDefinition) {
                            if (!config.table['columns']) {
                                config.table['columns'] = {};
                            }
                            if (Array.isArray(columnDefinition)) {
                                for (let j = 0; j < columnDefinition.length; j++) {
                                    evaluateFragmentColumn(columnDefinition[j], viewExtensionsOfPage, manifestKey, keyTab, config, extensionType, cellsFragmentName);
                                }
                            }
                            else {
                                evaluateFragmentColumn(columnDefinition, viewExtensionsOfPage, manifestKey, keyTab, config, extensionType, cellsFragmentName);
                            }
                        }
                    }
                }
            }
        }
    }
}
exports.addFragments = addFragments;
function createListReportConfig(manifest, flex, appSchema, fragments, logger) {
    // Initialize i18next
    i18n_1.initI18n();
    if (!manifest['sap.ui.generic.app']) {
        extensionLogger_1.log(logger, {
            severity: "error" /* Error */,
            message: i18next_1.default.t('NOFE'),
            location: {
                path: 'webapp/manifest.json',
                range: ['sap.ui.generic.app']
            }
        });
        return;
    }
    const v2Pages = manifest['sap.ui.generic.app'].pages;
    if (!v2Pages) {
        extensionLogger_1.log(logger, {
            severity: "error" /* Error */,
            message: i18next_1.default.t('NOPAGES', { appId: manifest['sap.app']['id'] }),
            location: {
                path: 'webapp/manifest.json',
                range: ['sap.ui.generic.app']
            }
        });
        return;
    }
    const config = { table: {}, filterBar: {} };
    const pageKeys = [];
    const v2Page = utils_1.findListReportPage(manifest['sap.ui.generic.app'].pages, pageKeys);
    if (!v2Page) {
        extensionLogger_1.log(logger, {
            severity: "error" /* Error */,
            message: i18next_1.default.t('NOALP'),
            location: {
                path: 'webapp/manifest.json',
                range: ['sap.ui.generic.app']
            }
        });
        return;
    }
    addSettings(config, manifest, appSchema, pageKeys[0], logger);
    flex.forEach((change) => {
        index_1.addFlex(config, change, appSchema, logger);
    });
    addFragments(config, manifest, v2Page, common_1.PageType.ListReport, fragments);
    return config;
}
exports.createListReportConfig = createListReportConfig;
