// @sap/ux-specification/dist/src/sync/v4/generate/objectPage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("../../../specification/common");
const common_2 = require("../../common");
const utils_1 = require("../utils/utils");
/**
 * Adds definitions for line items in object page sections to the app schema
 * @param facetId - key of the facet, as listed in FacetConfigs
 * @param appSchema - app schema in general
 * @param facet - the given facet from the UI annotations
 * @param facets - list of all facets
 */
function handleLineItem(facetId, appSchema, facet, facets) {
    if (facet.base !== 'LineItemFacet' || !facetId) {
        //no properties
        return;
    }
    const navigationProperty = facetId.split('::')[0];
    const schemaIdForOpSection = 'ObjectPageSectionTableV4<' + facetId + '>';
    appSchema['definitions'][schemaIdForOpSection] = JSON.parse(JSON.stringify(appSchema['definitions']['ObjectPageSectionTableV4']));
    if (facet.ID !== undefined) {
        appSchema['definitions'][schemaIdForOpSection].title = common_2.FacetTitlePrefix + facet.ID;
    }
    if (facet.Label) {
        appSchema['definitions'][schemaIdForOpSection].description = facet.Label;
    }
    appSchema['definitions'][schemaIdForOpSection].properties.table.$ref =
        '#/definitions/ObjectPageTableV4<' + facetId + '>';
    const idForTable = 'ObjectPageTableV4<' + facetId + '>';
    appSchema['definitions'][idForTable] = JSON.parse(JSON.stringify(appSchema['definitions']['ObjectPageTableV4']));
    appSchema['definitions'][idForTable].properties['columns'] = {
        $ref: '#/definitions/' + facetId
    };
    for (const facetKey in facets) {
        if (facets[facetKey].base === 'LineItemFacet' && facetKey.includes(navigationProperty)) {
            utils_1.addLineItemsType(appSchema, facets[facetKey]['lineItem'], facets[facetKey]['entityType'], facetId);
        }
    }
}
/**
 * Creates a section definition in app schema.
 * @param {FacetConfigs} facets - list of all facets.
 * @param {string} facetKey - facet key.
 * @param {object} sections - schema of current sections definition.
 * @param {object} appSchema - app specific schema that potentially gets enhanced.
 */
function addSection(facets, facetKey, sections, appSchema) {
    const facet = facets[facetKey];
    if (facet.base === 'CollectionFacet') {
        //handle collection facets
        sections['properties'][facetKey] = common_2.createSectionWithoutProperties(facet);
        sections['properties'][facetKey].properties = {
            subsections: {
                additionalProperties: false,
                type: 'object',
                properties: {}
            }
        };
        const subSections = sections['properties'][facetKey].properties.subsections;
        for (const key in facet['facets']) {
            addSection(facet['facets'], key, subSections, appSchema);
        }
    }
    else if (facet.base === 'LineItemFacet') {
        //handle reference facet (if comprising line item)
        sections['properties'][facetKey] = {
            $ref: '#/definitions/ObjectPageSectionTableV4<' + facetKey + '>'
        };
        handleLineItem(facetKey, appSchema, facet, facets);
    }
    else {
        sections['properties'][facetKey] = common_2.createSectionWithoutProperties(facet);
    }
}
/**
 * Derives the sections from the UI.Facets annotation and adds them to the schema
 * @param {object} appSchema - application specific JSON schema
 * @param {string} entityTypeName - current entity type
 * @param pages {SapUiAppPageV4[]} - list of (manifest) pages
 * @param {ConverterOutput} oDataServiceAVT - complete service information, as returned by annotation vocabularies tool
 */
function addSections(appSchema, entityTypeName, entitySetName, pages, oDataServiceAVT) {
    if (!pages) {
        delete appSchema['definitions']['ObjectPageSubSections'];
        return;
    }
    // Get facet annotations
    const facets = common_2.getObjectPageFacets(entityTypeName, oDataServiceAVT, common_1.FioriElementsVersion.v4);
    if (!facets) {
        delete appSchema['definitions']['ObjectPageSubSections'];
        return;
    }
    const sections = appSchema['definitions']['Sections'];
    Object.keys(pages).forEach((element) => {
        if (pages[element].options &&
            pages[element].options.settings &&
            pages[element].options.settings.entitySet === entitySetName &&
            pages[element].name === 'sap.fe.templates.ObjectPage') {
            for (const facetKey in facets) {
                addSection(facets, facetKey, sections, appSchema);
            }
        }
    });
    // Custom sections as part of sections
    if (appSchema['definitions']['CustomSections']) {
        sections.properties = Object.assign(Object.assign({}, sections.properties), appSchema['definitions']['CustomSections'].properties);
    }
}
/**
 * Adds the related facets keys 'enum' to the app schema
 * @param {object} schema - app specific schema that potentially gets enhanced
 * @param {string} entityTypeName - current entity type
 * @param {string} entitySetName - entity set of the actual (object) page
 * @param {SapUiAppPageV4[]} pages - list of all pages
 * @param {ConverterOutput} oDataServiceAVT - complete service information, as returned by annotation vocabularies tool
 */
function addRelatedFacetKeysType(schema, entityTypeName, entitySetName, pages, oDataServiceAVT) {
    Object.keys(pages).forEach((element) => {
        if (pages[element].options &&
            pages[element].options.settings &&
            pages[element].options.settings.entitySet === entitySetName &&
            pages[element].name === 'sap.fe.templates.ObjectPage') {
            // Find sections from annotation
            const facetSections = common_2.getObjectPageFacetSection(entityTypeName, oDataServiceAVT);
            // Find custom sections ids from manifest
            const page = pages[element];
            if (page.options &&
                page.options.settings &&
                page.options.settings.content &&
                page.options.settings.content.body &&
                page.options.settings.content.body.sections) {
                Object.keys(page.options.settings.content.body.sections).forEach(function (key) {
                    // Just in case - avoid duplications
                    if (!facetSections.find((section) => section.key === key)) {
                        facetSections.push({
                            key,
                            custom: true
                        });
                    }
                });
            }
            // Update schema definition
            common_2.addDefinitionForRelatedFacetKeys(schema, ['ObjectPageCustomSectionFragment'], facetSections, [
                'ID',
                'label',
                'key'
            ]);
        }
    });
}
/**
 * Generates an app specific schema out of the generic schema.
 * Generic types are replaced by information from the app specific annotations.
 * @param genericSchema  - generic JSON schema of an object page
 * @param entityTypeName - the base entity type of the given page
 * @param {ConverterOutput} oDataServiceAVT  - complete service information, as returned by annotation vocabularies tool
 * @param manifest - manifest.json of the app
 *
 * @returns the app specific JSON schema
 */
function generateObjectPageSchemaV4(genericSchema, entityTypeName, oDataServiceAVT, manifest) {
    const appSchema = JSON.parse(JSON.stringify(genericSchema));
    // Eliminate generic definitions
    delete appSchema['definitions']['GenericSections'];
    delete appSchema['definitions']['ObjectPageSubSections'];
    // Change reference to generated sections
    appSchema['properties']['sections'] = {
        $ref: '#/definitions/Sections'
    };
    //initialize sections
    appSchema['definitions']['Sections'] = {
        type: 'object',
        properties: {},
        additionalProperties: false
    };
    const entitySet = oDataServiceAVT.entitySets.find((es) => es.entityType.name === entityTypeName);
    // Add sections
    if (entitySet) {
        addSections(appSchema, entityTypeName, entitySet.name, manifest['sap.ui5'].routing.targets, oDataServiceAVT);
        // Custom Section facets
        addRelatedFacetKeysType(appSchema, entityTypeName, entitySet.name, manifest['sap.ui5'].routing.targets, oDataServiceAVT);
    }
    // delete generic definitions from schema
    delete appSchema['definitions']['ObjectPageSectionTableV4'];
    delete appSchema['definitions']['ObjectPageTableV4'];
    delete appSchema['definitions']['GenericColumns'];
    return appSchema;
}
exports.generateObjectPageSchemaV4 = generateObjectPageSchemaV4;
