// @sap/ux-specification/dist/src/sync/v4/import/pages/analyticalListPage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("../../../../specification/common");
const common_2 = require("../../../common");
const listReport_1 = require("./listReport");
const extensionLogger_1 = require("../../../../extensionLogger");
const i18next_1 = __importDefault(require("i18next"));
const i18n_1 = require("../../../../i18n/i18n");
const pages_1 = require("../../export/pages");
function createAnalyticalListPageConfig(manifest, jsonSchema, logger) {
    // Initialize i18next
    i18n_1.initI18n();
    if (!manifest['sap.ui5'] || !manifest['sap.ui5']['routing']) {
        extensionLogger_1.log(logger, {
            severity: "error" /* Error */,
            message: i18next_1.default.t('NOROUTING'),
            location: {
                path: 'webapp/manifest.json',
                range: ['sap.ui5']
            }
        });
        return;
    }
    const v4Pages = manifest['sap.ui5']['routing']['targets'];
    if (!v4Pages) {
        extensionLogger_1.log(logger, {
            severity: "error" /* Error */,
            message: i18next_1.default.t('NOTARGETS'),
            location: {
                path: 'webapp/manifest.json',
                range: ['sap.ui5', 'routing']
            }
        });
        return;
    }
    const config = new pages_1.AnalyticalListPage();
    Object.keys(v4Pages).forEach((key) => {
        const v4Page = v4Pages[key];
        if ('name' in v4Page && common_2.getPageTypeV4(v4Page.name) === common_1.PageTypeV4.AnalyticalListPage) {
            listReport_1.addSettings(config, manifest, jsonSchema, v4Page);
            return;
        }
    });
    return config;
}
exports.createAnalyticalListPageConfig = createAnalyticalListPageConfig;
