// @sap/ux-specification/dist/src/sync/v4/import/pages/listReport.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("../../../../specification/common");
const common_2 = require("../../../common");
const controls_1 = require("../../export/controls");
const common_3 = require("../common");
const i18n_1 = require("../../../../i18n/i18n");
const i18next_1 = __importDefault(require("i18next"));
const extensionLogger_1 = require("../../../../extensionLogger");
const pages_1 = require("../../export/pages");
const TableColumn_1 = require("../../export/controls/TableColumn");
function addSettings(listReportConfig, manifest, jsonSchema, page) {
    // Tranfer page level settings with sync rules
    common_3.transferSettingsOfObject(listReportConfig, manifest, jsonSchema, page.id, []);
    // Check table settings from target structure
    listReportConfig.table = new controls_1.Table();
    common_3.transferSettingsOfObject(listReportConfig.table, manifest, jsonSchema['definitions']['Table'], page.id, []);
    //transfer columns
    const manifestColumns = page.options &&
        page.options.settings &&
        page.options.settings.controlConfiguration &&
        page.options.settings.controlConfiguration['@com.sap.vocabularies.UI.v1.LineItem'] &&
        page.options.settings.controlConfiguration['@com.sap.vocabularies.UI.v1.LineItem']['columns'];
    if (manifestColumns) {
        listReportConfig.table.columns = {};
        for (const columnKey in manifestColumns) {
            const column = manifestColumns[columnKey];
            if (!column.template) {
                listReportConfig.table.columns[columnKey] = new TableColumn_1.TableColumn();
                common_3.transferSettingsOfObject(listReportConfig.table.columns[columnKey], manifest, jsonSchema['definitions']['TableColumn'], page.id, [columnKey]);
            }
        } //else: custom column ==> open ToDo
    }
    return listReportConfig;
}
exports.addSettings = addSettings;
function createListReportConfig(manifest, jsonSchema, logger) {
    // Initialize i18next
    i18n_1.initI18n();
    if (!manifest['sap.ui5'] || !manifest['sap.ui5']['routing']) {
        extensionLogger_1.log(logger, {
            severity: "error" /* Error */,
            message: i18next_1.default.t('NOROUTING'),
            location: {
                path: 'webapp/manifest.json',
                range: ['sap.ui5']
            }
        });
        return;
    }
    const v4Pages = manifest['sap.ui5']['routing']['targets'];
    if (!v4Pages) {
        extensionLogger_1.log(logger, {
            severity: "error" /* Error */,
            message: i18next_1.default.t('NOTARGETS'),
            location: {
                path: 'webapp/manifest.json',
                range: ['sap.ui5']
            }
        });
        return;
    }
    const config = new pages_1.ListReport();
    Object.keys(v4Pages).forEach((key) => {
        const v4Page = v4Pages[key];
        if ('name' in v4Page && common_2.getPageTypeV4(v4Page.name) === common_1.PageTypeV4.ListReport) {
            addSettings(config, manifest, jsonSchema, v4Page);
            return;
        }
    });
    return config;
}
exports.createListReportConfig = createListReportConfig;
