// @sap/ux-specification/dist/src/sync/v2/generate/listReport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("./utils");
const import_1 = require("../import");
function addFieldsType(appSchema, entityType) {
    const propertyNameArray = [], propertyObject = {};
    for (const key in entityType.entityProperties) {
        const propertyName = entityType.entityProperties[key].name;
        propertyNameArray.push(propertyName);
        propertyObject[propertyName] = {
            $ref: '#/definitions/FieldPath'
        };
    }
    appSchema['definitions']['EntityTypeProperties'] = {
        enum: propertyNameArray
    };
    appSchema['definitions']['FieldPath'].properties.path = {
        $ref: '#/definitions/EntityTypeProperties'
    };
    appSchema['definitions']['CreationFieldKeys'] = {
        type: 'object',
        properties: propertyObject,
        additionalProperties: false
    };
    appSchema['definitions']['ToolbarActions'].properties['createWithParameterDialog'].properties.fields = {
        $ref: '#/definitions/CreationFieldKeys'
    };
}
/**
 * Generates an app specific schema for the FE V2 Listreport from the generic schema.
 * Generic types are replaced by information from the app specific annotations.
 *
 * @param {object} genericSchema - generic JSON schema of a list report
 * @param {string} entityTypeName - the base entity type of the given page
 * @param {EntityType[]} entityTypes - list of all entity types
 * @param {object} manifest - manifest.json of the given application
 * @param {FileData[]} fragments - list of fragment files and thier content
 */
function generateListReportSchemaV2(genericSchema, entityTypeName, entityTypes, manifest, fragments) {
    const appSchema = JSON.parse(JSON.stringify(genericSchema));
    appSchema['properties']['table']['anyOf']['0']['$ref'] = '#/definitions/ResponsiveTableWithMultiSelect<LineItems>';
    appSchema['properties']['table']['anyOf']['1']['$ref'] = '#/definitions/ResponsiveTableWithInlineDelete<LineItems>';
    appSchema['properties']['table']['anyOf']['2']['$ref'] = '#/definitions/TreeTable<LineItems>';
    appSchema['properties']['table']['anyOf']['3']['$ref'] = '#/definitions/AnalyticalTable<LineItems>';
    appSchema['properties']['table']['anyOf']['4']['$ref'] = '#/definitions/GridTable<LineItems>';
    const responsiveTableMultiSelect = JSON.parse(JSON.stringify(appSchema['definitions']['ResponsiveTableWithMultiSelect<GenericColumns>']));
    const responsiveTableInlineDelete = JSON.parse(JSON.stringify(appSchema['definitions']['ResponsiveTableWithInlineDelete<GenericColumns>']));
    const treeTableDefinition = JSON.parse(JSON.stringify(appSchema['definitions']['TreeTable<GenericColumns>']));
    const analyticalTableDefinition = JSON.parse(JSON.stringify(appSchema['definitions']['AnalyticalTable<GenericColumns>']));
    const gridTableDefinition = JSON.parse(JSON.stringify(appSchema['definitions']['GridTable<GenericColumns>']));
    responsiveTableMultiSelect['properties']['columns'] = {
        $ref: '#/definitions/LineItems'
    };
    responsiveTableInlineDelete['properties']['columns'] = {
        $ref: '#/definitions/LineItems'
    };
    treeTableDefinition['properties']['columns'] = {
        $ref: '#/definitions/LineItems'
    };
    analyticalTableDefinition['properties']['columns'] = {
        $ref: '#/definitions/LineItems'
    };
    gridTableDefinition['properties']['columns'] = {
        $ref: '#/definitions/LineItems'
    };
    appSchema['definitions']['ResponsiveTableWithMultiSelect<LineItems>'] = responsiveTableMultiSelect;
    appSchema['definitions']['ResponsiveTableWithInlineDelete<LineItems>'] = responsiveTableInlineDelete;
    appSchema['definitions']['TreeTable<LineItems>'] = treeTableDefinition;
    appSchema['definitions']['AnalyticalTable<LineItems>'] = analyticalTableDefinition;
    appSchema['definitions']['GridTable<LineItems>'] = gridTableDefinition;
    delete appSchema['definitions']['ResponsiveTableWithMultiSelect<GenericColumns>'];
    delete appSchema['definitions']['ResponsiveTableWithInlineDelete<GenericColumns>'];
    delete appSchema['definitions']['TreeTable<GenericColumns>'];
    delete appSchema['definitions']['AnalyticalTable<GenericColumns>'];
    delete appSchema['definitions']['GridTable<GenericColumns>'];
    delete appSchema['definitions']['GenericColumns'];
    const entityType = entityTypes.find((et) => et.name === entityTypeName);
    //handle line items
    const lineItemAnnotation = entityType.annotations && entityType.annotations.UI && entityType.annotations.UI.LineItem;
    utils_1.addLineItemsType(appSchema, lineItemAnnotation, entityType);
    //add fields for createWithParameterDialog
    addFieldsType(appSchema, entityType);
    const pageKeys = [];
    const v2Page = import_1.findListReportPage(manifest['sap.ui.generic.app'].pages, pageKeys);
    utils_1.addListReportColumnExtensions(appSchema, entityType, v2Page, manifest, fragments);
    utils_1.checkGenerateRules(appSchema);
    return appSchema;
}
exports.generateListReportSchemaV2 = generateListReportSchemaV2;
