// @sap/ux-specification/dist/src/sync/v2/generate/objectPage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("../../common");
const utils_1 = require("./utils");
const utils_2 = require("../../common/utils");
const import_1 = require("../import");
const v2_1 = require("../../../specification/v2");
const common_2 = require("../../../specification/common");
/**
 * Method ensures that definitions for custom columns are created for passed entitySet.
 * @param {object} appSchema App schema in general
 * @param {object} manifest manifest.json of the app
 * @param {string} entitySetName Entity set name.
 * @param {string} entityTypeName Entity type name.
 * @param {ConverterOutput} oDataServiceAVT Complete service information, as returned by annotation vocabularies tool
 * @param {FileData[]} [fragments] Array with XML fragments.
 * @return {string} Name of custom columns definition.
 */
function ensureCustomColumnDefinitionExists(appSchema, manifest, entitySetName, entityTypeName, oDataServiceAVT, fragments) {
    const customColumnsDefinitionName = `${utils_1.CUSTOM_COLUMNS_DEFINITION}<${entitySetName}>`;
    if (appSchema['definitions'][customColumnsDefinitionName]) {
        // Definition is already created - we do not need create again
        return utils_1.CUSTOM_COLUMNS_DEFINITION;
    }
    // Create type specific 'TableCustomColumn' definition - use entity set as type.
    const customColumnDefinitionName = `${utils_1.CUSTOM_COLUMN_DEFINITION}<${entitySetName}>`;
    const customColumnDefinition = JSON.parse(JSON.stringify(appSchema['definitions']['TableCustomColumn']));
    appSchema['definitions'][customColumnDefinitionName] = customColumnDefinition;
    // Update definitions for type specific custom columns
    const pageKeys = [];
    const entityType = oDataServiceAVT.entityTypes.find((et) => et.name === entityTypeName);
    const v2Page = import_1.findObjectPage(manifest['sap.ui.generic.app'].pages, entitySetName, pageKeys);
    utils_1.addListReportColumnExtensions(appSchema, entityType, v2Page, manifest, fragments, customColumnsDefinitionName, customColumnDefinitionName);
    return customColumnsDefinitionName;
}
/**
 * Adds definitions for line items in object page sections to the app schema
 * @param {string} facetId - key of the facet, as listed in FacetConfigs
 * @param {object} appSchema - app schema in general
 * @param {FacetConfig} facet - the given facet from the UI annotations
 * @param {FacetConfigs} facets - list of all facets
 * @param {ConverterOutput} oDataServiceAVT - complete service information, as returned by annotation vocabularies tool
 * @param {object} manifest - manifest.json of the app
 * @param {FileData[]} [fragments] - array with XML fragments.
 */
function handleLineItem(facetId, appSchema, facet, facets, oDataServiceAVT, manifest, fragments) {
    if (facet.base !== 'LineItemFacet' || !facetId) {
        //no properties
        return;
    }
    const navigationProperty = facetId.split('::')[0];
    const schemaIdForOpSection = 'ObjectPageSectionTableV2<' + facetId + '>';
    appSchema['definitions'][schemaIdForOpSection] = {
        type: 'object',
        properties: {
            table: {
                anyOf: [
                    {
                        $ref: '#/definitions/ObjectPageResponsiveTableWithMultiSelect<' + facetId + '>'
                    },
                    {
                        $ref: '#/definitions/ObjectPageResponsiveTableWithInlineDelete<' + facetId + '>'
                    },
                    {
                        $ref: '#/definitions/ObjectPageAnalyticalTable<' + facetId + '>'
                    },
                    {
                        $ref: '#/definitions/ObjectPageGridTable<' + facetId + '>'
                    },
                    {
                        $ref: '#/definitions/ObjectPageTreeTable<' + facetId + '>'
                    }
                ]
            }
        },
        additionalProperties: false
    };
    if (facet.ID) {
        appSchema['definitions'][schemaIdForOpSection].title = common_1.FacetTitlePrefix + facet.ID;
    }
    if (facet.Label) {
        appSchema['definitions'][schemaIdForOpSection].description = facet.Label;
    }
    // Find target entity
    const targetEntity = facet.entityType
        ? oDataServiceAVT.entitySets.find((es) => es.entityType.name === facet.entityType.name || es.name === facet.entityType.name)
        : undefined;
    const schemaIdForOpResponsibleTableMultiSelect = 'ObjectPageResponsiveTableWithMultiSelect<' + facetId + '>';
    const schemaIdForOpResponsibleTableInlineDelete = 'ObjectPageResponsiveTableWithInlineDelete<' + facetId + '>';
    const schemaIdForOpAnalyticalTable = 'ObjectPageAnalyticalTable<' + facetId + '>';
    const schemaIdForOpGridTable = 'ObjectPageGridTable<' + facetId + '>';
    const schemaIdForOpTreeTable = 'ObjectPageTreeTable<' + facetId + '>';
    appSchema['definitions'][schemaIdForOpResponsibleTableMultiSelect] = JSON.parse(JSON.stringify(appSchema['definitions']['ObjectPageResponsiveTableWithMultiSelect<GenericColumns>']));
    appSchema['definitions'][schemaIdForOpResponsibleTableMultiSelect]['properties']['columns']['$ref'] =
        '#/definitions/' + facetId;
    appSchema['definitions'][schemaIdForOpResponsibleTableInlineDelete] = JSON.parse(JSON.stringify(appSchema['definitions']['ObjectPageResponsiveTableWithInlineDelete<GenericColumns>']));
    appSchema['definitions'][schemaIdForOpResponsibleTableInlineDelete]['properties']['columns']['$ref'] =
        '#/definitions/' + facetId;
    appSchema['definitions'][schemaIdForOpAnalyticalTable] = JSON.parse(JSON.stringify(appSchema['definitions']['ObjectPageAnalyticalTable<GenericColumns>']));
    appSchema['definitions'][schemaIdForOpAnalyticalTable]['properties']['columns']['$ref'] =
        '#/definitions/' + facetId;
    appSchema['definitions'][schemaIdForOpGridTable] = JSON.parse(JSON.stringify(appSchema['definitions']['ObjectPageGridTable<GenericColumns>']));
    appSchema['definitions'][schemaIdForOpGridTable]['properties']['columns']['$ref'] = '#/definitions/' + facetId;
    appSchema['definitions'][schemaIdForOpTreeTable] = JSON.parse(JSON.stringify(appSchema['definitions']['ObjectPageTreeTable<GenericColumns>']));
    appSchema['definitions'][schemaIdForOpTreeTable]['properties']['columns']['$ref'] = '#/definitions/' + facetId;
    let customColumnDefinitionKey;
    if (targetEntity) {
        // Make sure that custom columns definitions are created for passed entity
        customColumnDefinitionKey = ensureCustomColumnDefinitionExists(appSchema, manifest, targetEntity.name, facet.entityType.name, oDataServiceAVT, fragments);
    }
    for (const facetKey in facets) {
        if (facets[facetKey].base === 'LineItemFacet' && facetKey.includes(navigationProperty)) {
            utils_1.addLineItemsType(appSchema, facets[facetKey]['lineItem'], facets[facetKey]['entityType'], facetId, customColumnDefinitionKey);
        }
    }
}
/**
 * Creates a section definition in app schema.
 * @param {FacetConfigs} facets - list of all facets.
 * @param {string} facetKey - facet key.
 * @param {object} sections - schema of current sections definition.
 * @param {object} appSchema - app specific schema that potentially gets enhanced.
 * @param {object} manifest - manifest.json of the app.
 * @param {ConverterOutput} oDataServiceAVT - complete service information, as returned by annotation vocabularies tool.
 * @param {FileData[]} [fragments] - array with XML fragments.
 */
function addSection(facets, facetKey, sections, appSchema, manifest, oDataServiceAVT, fragments) {
    const facet = facets[facetKey];
    if (facet.base === 'CollectionFacet') {
        //handle collection facets
        const schemaIdForSection = facetKey.replace('@com.sap.vocabularies.UI.v1.Facets', facet.base);
        sections['properties'][schemaIdForSection] = common_1.createSectionWithoutProperties(facet);
        const collectionDefinition = 'ObjectPageSectionTableV2<' + schemaIdForSection + '>';
        sections['properties'][schemaIdForSection].properties = {
            subsections: {
                $ref: '#/definitions/' + collectionDefinition
            }
        };
        appSchema['definitions'][collectionDefinition] = common_1.createSectionWithoutProperties();
        for (const key in facet['facets']) {
            addSection(facet['facets'], key, appSchema['definitions'][collectionDefinition], appSchema, manifest, oDataServiceAVT, fragments);
        }
    }
    else if (facet.base === 'LineItemFacet') {
        //handle reference facet (if comprising line item)
        sections['properties'][facetKey] = {
            $ref: '#/definitions/ObjectPageSectionTableV2<' + facetKey + '>'
        };
        handleLineItem(facetKey, appSchema, facet, facets, oDataServiceAVT, manifest);
    }
    else if (facet.base === 'ChartFacet') {
        //handle reference facet (if comprising chart)
        const chartSettings = appSchema['definitions']['GenericSections']['additionalProperties']['anyOf'].find((element) => element['$ref'] === '#/definitions/ObjectPageSectionChartV2');
        sections['properties'][facetKey] = chartSettings;
        if (facet.ID !== undefined) {
            appSchema['definitions']['ObjectPageSectionChartV2'].title = common_1.FacetTitlePrefix + facet.ID;
            if (facet.Label) {
                appSchema['definitions']['ObjectPageSectionChartV2'].description = facet.Label;
            }
        }
    }
    else {
        // Other facets -> no properties
        sections['properties'][facetKey] = common_1.createSectionWithoutProperties(facet);
    }
}
/**
 * Adds the sections to the app schema
 * @param {object} appSchema - app specific schema that potentially gets enhanced
 * @param {string} entityTypeName - current entity type
 * @param {SapUiAppPagesV2} pages - list of (manifest) pages
 * @param {ConverterOutput} oDataServiceAVT - complete service information, as returned by annotation vocabularies tool
 */
function addSections(appSchema, entityTypeName, entitySetName, pages, oDataServiceAVT, manifest, fragments) {
    //initialize sections
    const sections = (appSchema['definitions']['Sections'] = {
        type: 'object',
        properties: {},
        additionalProperties: false
    });
    if (!pages) {
        return;
    }
    // Get facet annotations
    const facets = common_1.getObjectPageFacets(entityTypeName, oDataServiceAVT, common_2.FioriElementsVersion.v2);
    if (!facets || Object.keys(facets).length === 0) {
        return;
    }
    // Loop on all pages, look for the right entity set
    for (const key in pages) {
        const element = pages[key];
        if (entitySetName.includes(element.entitySet)) {
            // Loop on all facets (of the given entitySet)
            for (const facetKey in facets) {
                addSection(facets, facetKey, sections, appSchema, manifest, oDataServiceAVT, fragments);
            }
        }
        else {
            addSections(appSchema, entityTypeName, entitySetName, element.pages, oDataServiceAVT, manifest, fragments);
        }
    }
    // Add custom sections
    for (const name in appSchema['definitions']['CustomSections'].properties) {
        sections.properties[name] = appSchema['definitions']['CustomSections'].properties[name];
    }
}
/**
 * Add Object Page Header Action Buttons to app-specific schema
 *
 * @param {object} appSchema Schema of the app
 * @param {string} entityTypeName - current entity type
 * @param {EntityType[]} entityTypes
 */
function addHeaderActions(appSchema, entityTypeName, oDataServiceAVT) {
    appSchema['definitions']['ObjectPageHeader'] = JSON.parse(JSON.stringify(appSchema['definitions']['ObjectPageHeader<GenericActions>']));
    delete appSchema['definitions']['ObjectPageHeader']['properties']['actions'];
    const entityType = oDataServiceAVT.entityTypes.find((et) => et.name === entityTypeName);
    if (!entityType) {
        return;
    }
    const alias = common_1.findAlias('com.sap.vocabularies.UI.v1', oDataServiceAVT);
    const actionAnnotation = alias &&
        entityType.annotations &&
        entityType.annotations[alias] &&
        entityType.annotations[alias].Identification;
    if (actionAnnotation) {
        // Add action to configuration
        appSchema['definitions']['ObjectPageHeader'].properties['actions'] = {
            $ref: '#/definitions/HeaderActions'
        };
        const headerActions = (appSchema['definitions']['HeaderActions'] = {
            type: 'object',
            properties: {},
            additionalProperties: false
        });
        actionAnnotation.forEach((item) => {
            if (!item['Determining'] && item.$Type === 'com.sap.vocabularies.UI.v1.DataFieldForAction') {
                const actionName = item.Action.split('.')[1].split('/');
                const entityName = entityType.fullyQualifiedName.split('.');
                const actionId = entityName[0] + '.' + actionName[0] + '::' + actionName[1];
                headerActions.properties[actionId] = {
                    $ref: '#/definitions/HeaderAction'
                };
            }
        });
    }
}
/**
 * Adds the related facets keys 'enum' to the app schema
 * @param schema - app specific schema that potentially gets enhanced
 * @param entityType - current entity type
 * @param {ConverterOutput} oDataServiceAVT - complete service information, as returned by annotation vocabularies tool
 */
function addRelatedFacetKeysType(schema, entityType, oDataServiceAVT) {
    // Find sections from annotation
    // And create new definition in schema as enum with description
    const facetSections = common_1.getObjectPageFacetSection(entityType, oDataServiceAVT);
    if (facetSections) {
        utils_2.addDefinitionForRelatedFacetKeys(schema, ['ObjectPageCustomSectionFragment', 'ObjectPageCustomSectionView'], facetSections, ['ID', 'key']);
    }
}
/**
 * Generates an app specific schema out of the generic schema.
 * Generic types are replaced by information from the app specific annotations.
 *
 * @param genericSchema - generic JSON schema of an object page
 * @param entityTypeName - the base entity type of the given page
 * @param annotations - list of all annotation files (file content plus URI as identifier)
 * @param {ConverterOutput} oDataServiceAVT - complete service information, as returned by annotation vocabularies tool
 * @param manifest - manifest.json of the app
 * @param {FileData[]} [fragments] - array with XML fragments.
 *
 * @returns the app specific JSON schema
 */
function generateObjectPageSchemaV2(genericSchema, entityTypeName, oDataServiceAVT, manifest, fragments) {
    const appSchema = JSON.parse(JSON.stringify(genericSchema));
    // Custom sections as part of sections
    const additionalSections = {};
    for (const name in appSchema['definitions']['CustomSections'].properties) {
        additionalSections[name] = appSchema['definitions']['CustomSections'].properties[name];
    }
    // Change reference to generated sections
    appSchema['properties']['sections'] = {
        $ref: '#/definitions/Sections'
    };
    // Change reference to header actions
    appSchema['properties']['header'] = {
        $ref: '#/definitions/ObjectPageHeader'
    };
    const entitySet = oDataServiceAVT.entitySets &&
        oDataServiceAVT.entitySets.find((es) => es.entityType.name === entityTypeName || es.name === entityTypeName);
    // Add sections
    if (entitySet) {
        addSections(appSchema, entityTypeName, entitySet.name, manifest['sap.ui.generic.app'].pages, oDataServiceAVT, manifest, fragments);
        delete appSchema['definitions']['ObjectPageTable<GenericColumns>'];
        // Add header actions
        addHeaderActions(appSchema, entityTypeName, oDataServiceAVT);
        delete appSchema['definitions']['GenericActions'];
        delete appSchema['definitions']['ObjectPageHeader<GenericActions>'];
        // Custom section facets
        addRelatedFacetKeysType(appSchema, entityTypeName, oDataServiceAVT);
    }
    // Eliminate generic Sections
    delete appSchema['definitions']['GenericSections'];
    delete appSchema['definitions']['ObjectPageSectionTableV2<GenericColumns>'];
    delete appSchema['definitions']['GenericColumns'];
    delete appSchema['definitions']['ObjectPageSubSections'];
    delete appSchema['definitions']['ObjectPageResponsiveTableWithMultiSelect<GenericColumns>'];
    delete appSchema['definitions']['ObjectPageResponsiveTableWithInlineDelete<GenericColumns>'];
    delete appSchema['definitions']['ObjectPageAnalyticalTable<GenericColumns>'];
    delete appSchema['definitions']['ObjectPageTreeTable<GenericColumns>'];
    delete appSchema['definitions']['ObjectPageGridTable<GenericColumns>'];
    // Remove Custom Column non existing extensions - TreeTableColumnsExtension does not exist in ALP
    utils_1.removeElementsFromEnum(appSchema['definitions']['TableColumnExtensionTypeV2']['enum'], [
        v2_1.TableColumnExtensionTypeV2.TreeTableColumnsExtension
    ]);
    if (!appSchema['definitions']['Sections']) {
        if (appSchema['definitions']['RelatedFacetKeys'] &&
            appSchema['definitions']['RelatedFacetKeys']['oneOf'].length) {
            // If related facets exists - then we need sections definition
            appSchema['definitions']['Sections'] = {
                type: 'object',
                properties: additionalSections,
                additionalProperties: false
            };
        }
        else {
            delete appSchema['properties']['sections'];
        }
    }
    utils_1.checkGenerateRules(appSchema);
    return appSchema;
}
exports.generateObjectPageSchemaV2 = generateObjectPageSchemaV2;
